KAFKA-4123: Queryable State returning null for key before all stores in instance have been initialized

dguy opened the pull request:
Mark the store as open after the DB has been restored from the changelog.
Only add the store to the map in ProcessorStateManager post restore.
Make RocksDBWindowStore.Segment override openDB(..) as it needs to mark the Segment as open.
Throw InvalidStateStoreException if any stores in a KafkaStreams instance are not available.

dguy committed:
only mark store as open after it has been initialized. Add store to map in state manager post initiliazation

dguy commented:
This was reported on the dev list in an email in relation to Queryable State read guarantees. It was possible that on restore of a `StateStore`, i.e, on rebalance or app startup, that QS could start reading from the `StateStore` before it had been completely restored. There were 2 reasons this was possible:
1.  a `RocksDBStore` was marked as open in  `openDB(..)` which is called before the state has been restored. This was because we don't call `init` on `RocksDBWindowStore.Segment`
2. In `ProcessorStateManager` the store was added to the `stores` map before the state had been restored.

Unfortunately i couldn't manage to write a reliable failing test for this as it is really down to timing of the calls and multiple threads.

Please have a look @enothereska @guozhangwang

enothereska commented:
LGTM

dguy commented:
There is another issue here:
 In the case of multiple partitions and the tasks are being initialized it is always possible that 1 or more StateStores will be intialized before the complete set of stores in the Streams Instance are initialized. Currently when this happens a query can return null because it will look in the existing initialized stores. However, the key may exist in one of the non-initialized instances. We need to wait for all Stores in the instance to be initialized before allowing queries to progress

dguy committed:
only allow querying of stores when all stores are available

dguy commented:
I've now fixed the issue. Made it so that an exceptions is raised when all stores aren't available. Added an integration test (that was failing prior to the change). @enothereska @guozhangwang  please review

dguy commented:
FYI  - test failure is in core, so unrelated to this change

enothereska commented:
LGTM, good test.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -105,6 +105,7 @@
 
     private Map<TopicPartition, List<ConsumerRecord<byte[], byte[]>>> standbyRecords;
     private boolean processStandbyRecords = false;
+    private AtomicBoolean stateStoresAvailable = new AtomicBoolean(false);
```
nit: could we rename this boolean to `initialized`? And then when checking this flag is not true:

```
throw new InvalidStateStoreException("Store: " + storeName + " may not currently available, since the hosted thread is not (re-)initialized yet");
```

Also we may need to update the `ReadOnlyKeyValueStore` and `ReadOnlyWindowStore` about those possible thrown exceptions and how to handle them (e.g. backoff and retry).

guozhangwang commented:
Left a comment about updating javadoc on possible thrown exceptions for public APIs. Otherwise LGTM.

dguy committed:
move InvalidStateStoreException to errors package. Add comment about backoff and retry. Update ReadOnlyKeyValueStore and ReadOnlyWindowStore with throws InvalidStateStoreException

dguy commented:
@guozhangwang I've changed the variable in `StreamThread` to `initialized` as suggested and also updated the exception thrown.
Also, i moved the `InvalidStateStoreException` to `org.apache.kafka.streams.errors` as that seems like where it should be. I updated the javadoc of `InvalidStateStoreException` to mention about the transient nature of these exceptions and that it is valid to backoff and retry.
I also added `throws InvalidStateStoreException` to all `ReadOnlyKeyValueStore` and `ReadOnlyWindowStore` javadocs

dguy committed:
merge trunk

guozhangwang commented:
Created https://issues.apache.org/jira/browse/KAFKA-4143 for the transient failure in Jenkins.
